Add TLS version changes to deprecation checks

tvernum opened the pull request:
This adds a warning to the deprecations API for any SSL contexts that
rely on the default `supported_protocols` list. This list will change
in ES 7.0 and will no longer include TLS 1.0 by default.

Relates: #37512

tvernum committed:
Add TLS version changes to deprecation checks

This adds a warning to the deprecations API for any SSL contexts that
rely on the default `supported_protocols` list. This list will change
in ES 7.0 and will no longer include TLS 1.0 by default.

tvernum committed:
Merge branch '6.x' into deprecation-api-tls10

elasticmachine commented:
Pinging @elastic/es-security

tvernum commented:
@gwbrown @jaymode 
We have a problem here (and as best I can tell #36847 has the same issue).
It looks like the `Settings` that get passed into the deprecation check are _filtered_ so none of the SSL settings are there.

A config like this:
```
xpack.security.transport.ssl.enabled: true
xpack.security.transport.ssl.supported_protocols: [ "TLSv1.2", "TLSv1.1", "TLSv1" ]
xpack.ssl.certificate_authorities: [ "tls/ca/ca.crt" ]
xpack.ssl.certificate: tls/server/server.crt
xpack.ssl.key: tls/server/server.key
```
should generate 2 deprecation warnings:
1. due to the use of `xpack.ssl.*` which will be removed
2. the use of `xpack.ssl.*` with the default protocols.

It should _not_ generate a default protocols warning for `xpack.security.transport.ssl`, as the protocols are explicitly configured.

However, it does the exact opposite. The deprecation check sees filtered settings, so it only gets:
```
xpack.security.transport.ssl.enabled: true
```
and it doesn't generate any `xpack.ssl.*` warnings, but does generate one for `xpack.security.transport.ssl`.

@gwbrown Is it intentional that the deprecation checks run on filtered settings? Any thoughts on how to resolve this?

tvernum commented:
This is because the deprecation checks rely on `NodeInfo` and NodeService always filters the returned settings (which makes sense, that's the point of the filter) but it means we can't write node level deprecation checks that depend on filtered settings.
https://github.com/elastic/elasticsearch/blob/0227260/server/src/main/java/org/elasticsearch/node/NodeService.java#L90

tvernum committed:
Explicitly set protocols in rest test

We want the default deprecation rest tests to run without any
deprecations. This means we need to configure "supported_protocols"
for transport.ssl

Note: This doesn't currently work because supported_protocols is a
filtered setting, and filtered settings are not available in
deprecation checks

gwbrown commented:
I've opened https://github.com/elastic/elasticsearch/issues/37845 to track the problem, as it's not confined to this check.

jaymode commented:
Nice catch Tim.

jaymode reviewed (approved):
The changes here LGTM. I guess merging depends on what we're going to do with these that need filtered settings

AthenaEryma committed:
Merge branch '6.x' into pr/37793

AthenaEryma committed:
Add TLS settings to client build.gradle

AthenaEryma committed:
Unused imports

gwbrown commented:
@tvernum The fixes for the bug discussed above are merged to master - I went ahead and merged master into this PR and made the required changes on a branch, you can take a look here: https://github.com/elastic/elasticsearch/compare/6.x...gwbrown:pr/37793

tvernum committed:
Merge branch '6.x' into deprecation-api-tls10

tvernum committed:
Add deprecation check for https monitoring

A https monitoring exporter may not have any "ssl.*" settings, but
should still issue a deprecation warning as target cluster (or
infrastructure in between) may require TLSv1.0

tvernum commented:
Thanks @gwbrown I'ved merged in your changes.

@jaymode I've also added a change to explicitly check for "https" monitoring exporters that don't have `ssl.supported_protocols` (similar to LDAP & SAML realms).

tvernum committed:
Merge branch '6.x' into deprecation-api-tls10

tvernum committed:
Fix tests due to merge
